### PR TITLE
qwen4_exp: load checkpoints with pre-shifted RMSNorm gains

### LINKED ISSUE #2041: qwen4_exp: Qwen3.8-Flash-Next MLX 4-bit checkpoint loads cleanly on main but generates garbage
`Vontra/Qwen3.8-Flash-Next-MLX-4bit` loads cleanly on `main` — every weight maps, no warnings — but generation is deterministic garbage. The same checkpoint files produce coherent text on the superseded #2028 branch, so I don't think it's the checkpoint.

Raising it because the failure is silent: nothing in the load path complains, and `mlx_vlm/tests/test_qwen4_exp.py` passes 8/8.

## Repro

```sh
pip install "git+https://github.com/Blaizzy/mlx-vlm@main"   # c2a67b63
hf download Vontra/Qwen3.8-Flash-Next-MLX-4bit --local-dir Q38FN-4bit

python -m mlx_vlm.generate --model Q38FN-4bit \
  --prompt "Write a short paragraph explaining what Gated DeltaNet is and why it helps long-context inference." \
  --max-tokens 128 --temperature 0.0 --verbose
```

Output:

```
  
4<|im_start|>
###

Prompt: 31 tokens, 1.070 tokens-per-sec
Generation: 7 tokens, 33.636 tokens-per-sec
Peak memory: 111.835 GB
```

Seven tokens and it stops. Byte-identical across runs and across `505267ca` / `c2a67b63`, so it's deterministic rather than a sampling artifact.

For contrast, the same directory on `JJJYmmm:add_qwen4exp` (#2028) — plus two small compat patches, see below — gives:

> Gated DeltaNet is an advanced recurrent neural network architecture that integrates the delta rule (a fast, parallelizable update mechanism) with a gating mechanism to manage information flow within the hidden state. It helps long-context inference by enabling efficient, linear-time processing of sequences, avoiding the quadratic computational complexity and memory overhead associated with traditional Transformers' self-attention mechanisms. […]

at 33.9 tok/s decode, 658 tok/s prefill on a 2,585-token prompt.

## What I ruled out

- **Checkpoint integrity** — 22/22 shards, 3,671 tensors, on-disk bytes exactly equal `metadata.total_size` (111,578,339,151).
- **Mixed precision hiding in the checkpoint** — all 998 quantized modules have `weight_cols / scales_cols == 4.0`, i.e. uniformly 4-bit at group 32, matching `config.json`'s `{"group_size": 32, "bits": 4, "mode": "affine"}`. No per-module exceptions.
- **Weight mapping** — `load_weights` is strict here and raised nothing, so the key sets match exactly after `sanitize`.
- **n-gram hash constants** — `_build_layer_multipliers` and the `_find_nth_prime_after` head tables on `main` produce values identical to the #2028 branch: multipliers `[23703573157769, 20109073645365, 8052911324071]`, head sizes `[20000003, 20000023, 20000033, 20000047, …]`, offsets `[0, 20000003, 40000026, 60000059, …]`, total 320,001,446 → padded 320,001,536 = 128 × 2,500,012, which is exactly the shard layout in the checkpoint.
- **`conv1d` axis handling** — all 37 `conv1d.weight` tensors are `(10240, 4, 1)`, so `shape[-1] != 1` is false and the `moveaxis` in `sanitize` correctly no-ops.
- **Stale checkout** — reproduced on `c2a67b63` with `test_qwen4_exp.py` from the same commit passing 8/8. (An earlier 1/8 failure on my side was my own version skew, not a real failure.)

So the weights are all present, correctly shaped, correctly dequantized as far as I can tell, and the derived constants agree — which points at the forward pass rather than loading.

## Question

Is this checkpoint layout expected to work on `main`? It was converted before #2032 landed, with the #2028 branch's converter. If community conversions need redoing against `main`'s converter, that's worth a line in `models/qwen4_exp/README.md` — several 4-bit/8-bit MLX conversions of this model were published on the Hub in the hours around the merge, and this failure mode gives users no hint that the checkpoint is the problem.

Happy to run diagnostics on my side — I have the model resident and can dump intermediate activations, per-layer norms, or first-token logits against the #2028 implementation for the same input if that would help localize it.

## Environment

- M4 Max, 128 GB, macOS 26.5.2, `iogpu.wired_limit_mb=122880`
- `mlx 0.32.2`, `mlx-vlm` @ `c2a67b63`, `transformers 5.16.1`, Python 3.11.14
- Peak memory 111.8 GB on `main` vs 135.9 GB on #2028 — the sharded n-gram embedding is a real improvement, worth keeping

<sub>Two patches were needed for #2028 to read this checkpoint at all, in case they're useful context: its `config.py` rejects `layer_types: "full_attention"` (`main` already maps it to `qwen_sparse_attention`), and its `merge_ngram_embedding_shards` only matches `.shard_N.weight`, so quantized shards leave `.scales`/`.biases` unmatched (`main`'s `ShardedEmbedding` handles this correctly).</sub>

### PR DESCRIPTION
**Pull request by Lazarus-931**
Fixes #2041.

`Qwen4ExpRMSNorm` applies `1 + w` to gains the released checkpoint stores centered at zero, matching upstream `Qwen4ExpTextRMSNorm`, but MLX conversions published before #2032 folded that `+1` into their saved weights — so loading one applies the offset twice and inflates every gain (~2.6x on the hyper-connection norms), which is the noise reported in the issue. Because such a checkpoint is otherwise identical to a correct one in keys, shapes and dtypes, `sanitize()` now tells them apart by value — released gains average ~+0.22, pre-shifted ones ~+1.15 — and subtracts the offset back out with a warning pointing at re-conversion. The same correction is applied to the MTP draft head, which shares `Qwen4ExpRMSNorm` and had the same defect, and `text_config.preshifted_norm_weights` overrides the detection either way.

Verified end-to-end on `Vontra/Qwen3.8-Flash-Next-MLX-4bit` (gain center +1.22, 148 tensors corrected): `main` stops after 7 tokens of noise, matching the issue exactly, the fix generates coherent text to the full 128-token limit, and forcing the detection off reproduces the 7-token failure.

Surveying the Hub, 8 of 11 published MLX conversions carry the doubled offset, and there was no canonical conversion to point people at — so I made one:

link to correct model: https://huggingface.co/AlazarM/Qwen3.8-Flash-Next-4bit

Converted from `Qwen/Qwen3.8-Flash-Next` with this repo's own converter on `main`, with norm gains verified bit-identical to the source (148/148) and a byte-exact source integrity check.

### CODE REVIEW

**Comment by lucasnewman** [comment]
Seems like we should just upload correct conversions to `mlx-community`? Supporting every busted conversion with model-specific code seems like a mistake.

**Comment by Lazarus-931** [comment]
> Seems like we should just upload correct conversions to `mlx-community`? Supporting every busted conversion with model-specific code seems like a mistake.

yeah I think I'll do a proper conversion, mlx-community doesn't seem to have qwen3.8-next
